feat: KYZON Space Conferencing App

nangelina opened the pull request:
## What does this PR do?

Your friends at [KYZON Solutions](https://kyzonsolutions.com/space) are excited to introduce the **KYZON Space** conferencing app for Cal.com.

When installed, users can set **KYZON Space** as the location for their event types. Each booking will automatically generate a unique Space call link, ensuring a fresh room for every meeting.

The app integrates with the KYZON API via OAuth, storing an access token for each user. This token is then used to request and attach new Space call links to bookings.

To build this integration, I followed the [How to Build an App guide](https://cal.com/docs/how-to-guides/app-store-and-integrations/how-to-build-an-app#steps-in-creating-an-app) and drew on patterns from existing video conferencing apps (e.g. Zoom, Teams, Jelly) to implement dynamic link generation.

Although the contributor docs generally recommend not committing lock files, based on [this previous PR](https://github.com/calcom/cal.com/pull/13929), I included it here since we’re introducing a new package to the monorepo.

## Visual Demo

#### Video Demo:

https://github.com/user-attachments/assets/b9b42e76-6993-4786-92b9-1ec26f1daaa5

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). N/A
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

Please contact me at [angelina@kyzonsolutions.com](mailto:angelina@kyzonsolutions.com) for the `client_id`, `client_secret`, and `api_key` that you'll need to set to test this change. 

Currently the only OAuth callback URLs we've been configured to accept are:
- `https://app.cal.com/api/integrations/kyzonspacevideo/callback`, and 
- `https://cal.com/api/integrations/kyzonspacevideo/callback`.

If you plan to test this with different OAuth callback URL, please let me know so that I can configure our server to accept those requests!

<!-- devin-review-badge-begin -->

---

<a href="https://app.devin.ai/review/calcom/cal.com/pull/23486" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    <img src="https://static.devin.ai/assets/gh-open-in-devin-review-light.svg?v=1" alt="Open with Devin">
  </picture>
</a>
<!-- devin-review-badge-end -->

nangelina committed:
initial generation with event-type-location-video-static template

nangelina committed:
implement kyzon-space conferencing app. renamed type from kyzon-space_conferencing to kyzon-space_video due to the delete meeting api only being called for types ending with _video.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=23486) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=23486) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=23486) it.</sub>

nangelina committed:
add explicit return type to getKyzonAppKeys to enforce the parsed type at compile time.

nangelina committed:
select only required fields from prisma

nangelina committed:
fix "X-API-Key" header casing inconsistency

nangelina committed:
use WEBAPP_URL const instead of process.env.NEXT_PUBLIC_WEBAPP_URL for redirect_uri to be consistent

nangelina committed:
await setDefaultConferencingApp before redirecting to avoid race conditions

nangelina committed:
Rename deleteMeeting param for clarity: uid → meetingId

nangelina committed:
make sure cal's event start and end times are converted to ISO string format to guarantee UTC.

nangelina committed:
refactor add.ts handler function to include NextApiResponse and return proper JSON responses

nangelina committed:
remove logging errors to console in callback.ts to avoid leaking secrets

nangelina committed:
Refactor kyzonCredentialKeySchema: Tighten the schema and coerce IDs to avoid runtime surprises

nangelina committed:
Refactor getKyzonCredentialKey: Adjust expiry_date calculation to account for clock-skew and add runtime validation

nangelina committed:
Guard missing refresh_token before calling provider

nangelina committed:
Instead of logging the whole error object, only log select fields to avoid dumping secrets

nangelina committed:
Add a timeout to kyzonAxiosInstance to ensure all external API calls fail fast and avoid blocking the app

nangelina committed:
Harden expiry check against malformed tokens

Fail safe to “expired” if expiry_date is missing or invalid.

nangelina committed:
Use URLSearchParams instead of legacy querystring.stringify.

URLSearchParams handles RFC 3986 encoding more predictably for OAuth params.

nangelina committed:
Instead of logging the whole error object, only log select fields to avoid dumping secrets

nangelina commented on packages/app-store/kyzon-space/config.json:
```diff
@@ -0,0 +1,24 @@
+  "/*": "Don't modify slug - If required, do it using cli edit command",
+  "name": "KYZON Space",
+  "slug": "kyzon-space",
+  "type": "kyzon-space_video",
```
In the code for `getVideoAdaptors` in videoClient.ts, line 28 contains
```ts
const appName = cred.type.split("_").join(""); // Transform `zoom_video` to `zoomvideo`;
```
so although my type is "kyzon-space_video", its expecting an appName of "kyzon-spacevideo"

nangelina commented on packages/app-store/kyzon-space/lib/axios.ts:
```diff
@@ -0,0 +1,7 @@
+
+export const kyzonAxiosInstance = axios.create({
+  baseURL: kyzonBaseUrl,
+});
```
I don't have separate base URLs for my oauth routes vs my other api routes.

nangelina committed:
Previously, in 1e36d6658e78c11f59b81fa4c483b946a4f5b74f, we applied a 60s skew when computing expiry_date.

However isTokenExpired has another 5m buffer. That double-buffer caused short-lived tokens to appear expired immediately, leading to perpetual refreshes.

I’ve simplified this by removing the 5min buffer. Now expiry_date always represents the “safe-to-use-until” time, and isTokenExpired only needs to do a direct comparison against Date.now(). This keeps the logic clear and avoids stacking buffers in multiple places.

nangelina commented on packages/app-store/kyzon-space/lib/VideoApiAdapter.ts:
```diff
@@ -0,0 +1,244 @@
+import { kyzonAxiosInstance } from "./axios";
+import { refreshKyzonToken, isTokenExpired } from "./tokenManager";
+
+const KyzonVideoApiAdapter = (credential: CredentialPayload): VideoApiAdapter => {
+  const getRefreshedKey = async (): Promise<KyzonCredentialKey> => {
+    let key = kyzonCredentialKeySchema.parse(credential.key);
+
+    // Check if token needs refresh and has refresh capability
+    if (isTokenExpired(key)) {
+      const refreshedToken = await refreshKyzonToken(credential.id);
+      if (refreshedToken) {
+        key = refreshedToken;
+      }
+    }
+
+    return key;
+  };
+
```
Thanks for the suggestion! I'm not too sure if I can use this suggestion as is, as my current code already wraps my axios requests in a try catch, so your withAuth function wouldn't catch the 401s

nangelina committed:
add descriptive user-agent to kyzonAxiosInstance, so KYZON can whitelist its requests

nangelina committed:
implement wrapper function: authenticatedRequest. this will retry a function once if it errors with 401

nangelina committed:
De-duplicate concurrent refreshes per credential.

Prevent thundering-herd refresh when many requests hit an expired token simultaneously.

nangelina commented on packages/app-store/kyzon-space/lib/tokenManager.ts:
```diff
@@ -0,0 +1,99 @@
+
+async function _refreshKyzonToken(credentialId: number): Promise<KyzonCredentialKey | null> {
+  try {
+    const credential = await prisma.credential.findUnique({
+      where: { id: credentialId },
+      select: {
+        key: true,
+      },
+    });
+
```
This function requires getting the credentials, as they are needed to refresh the app's tokens. This is not a tRPC endpoints or an API, and we dont return these credentials anyway.

nangelina committed:
change config of kyzon space call to have wait room enabled

nangelina committed:
fix: rename kyzon-space app directory to kyzonspacevideo

- renamed app directory from `kyzon-space` → `kyzonspacevideo` because `video.adapters.generated.ts` generates keys from the directory name.
- otherwise, `getVideoAdapters()` in `videoClient.ts` will not be able to resolve `videoAdapterImport`, as it gets the key of `VideoApiAdapterMap` from the "type" of the app, which is kyzonspace_video.
- the "_video" suffix is required in the type, as otherwise the `deleteMeeting` method wouldn't be called, due to code that checks for whether the app type ends with "_video".
- update slug to `kyzonspacevideo` as well, since assets (e.g. `logo.svg`) are fetched using that path.

nangelina committed:
change config of kyzon space call to have wait room enabled

Devanshusharma2005 committed:
Merge branch 'main' into angelina/kyzon-space-app

Devanshusharma2005 commented on packages/app-store/kyzonspacevideo/DESCRIPTION.md:
```diff
@@ -0,0 +1,9 @@
+  - 4.png
+---
+
+{DESCRIPTION}
```
we don't need to add description . You can remove this.

Devanshusharma2005 reviewed (changes requested):
lets add tests also to verify this.

nangelina committed:
better user-facing error messages

also handle when updateMeeting 404s, now we will create the meeting if it doesnt exist when updating it

nangelina committed:
write tests

nangelina commented:
@Devanshusharma2005 I think I've covered all the points you've raised so far in your comments. Thank you for the PR review!

Devanshusharma2005 committed:
Merge branch 'main' into angelina/kyzon-space-app

Devanshusharma2005 committed:
Merge branch 'main' into angelina/kyzon-space-app

Devanshusharma2005 committed:
Merge branch 'main' into angelina/kyzon-space-app

nangelina committed:
Merge remote-tracking branch 'calcom/main' into angelina/kyzon-space-app

Devanshusharma2005 committed:
Merge branch 'main' into angelina/kyzon-space-app

romitg2 committed:
Merge branch 'main' into angelina/kyzon-space-app

romitg2 committed:
Merge remote-tracking branch 'upstream/main' into angelina/kyzon-space-app

romitg2 committed:
Merge remote-tracking branch 'upstream/main' into angelina/kyzon-space-app

romitg2 commented on packages/app-store/kyzonspacevideo/config.json:
```diff
@@ -0,0 +1,24 @@
+  "appData": {
+    "location": {
+      "type": "integrations:{SLUG}_video",
+      "label": "{TITLE}",
```
can you please update value of `{SLUG}` and `{TITLE}`

nangelina replied:
Hi, I'm not sure what I should be changing `{SLUG}` and `{TITLE}` to; I've checked how the other conferencing apps do it and I've seen some use `{SLUG}` and `{TITLE}`, and some have hardcoded their app names into it.

I did test it in local dev and it currently works with `{SLUG}` and `{TITLE}`.

Although if you prefer I also change `{SLUG}` and `{TITLE}` to the hardcoded values then I'm happy to do so as well. :)

nangelina committed:
fix: remove logging sensitive information (full response body & authorisation code)

nangelina committed:
fix: update token_type schema to allow case-insensitive 'Bearer' string

nangelina committed:
fix: prevent null/undefined coercing to strings for user_id and team_id

nangelina committed:
fix: update type to 'kyzonspacevideo_video'
Reverts the workaround from b9f5b5e as the underlying getVideoAdapters bug is resolved.

romitg2 committed:
Merge branch 'main' into angelina/kyzon-space-app

romitg2 committed:
Merge branch 'main' into angelina/kyzon-space-app

romitg2 reviewed (changes requested):
@nangelina can you please [fix failing tests](https://github.com/calcom/cal.com/actions/runs/22320886508/job/64578934564?pr=23486) also can you check 2 review comments from devin ? Thank you

nangelina committed:
fix error handling to be consistent with the other two error paths: redirect when state provides a return URL, otherwise fall back to the JSON error response. The two separate checks are also collapsed into one since the redirect logic is identical for both cases.

nangelina committed:
cache refreshed token in KyzonVideoApiAdapter to avoid redundant OAuth refreshes

nangelina committed:
fix failing tests

romitg2 committed:
Merge branch 'main' into angelina/kyzon-space-app

romitg2 committed:
feat: update seed app store script

romitg2 committed:
Merge branch 'main' into angelina/kyzon-space-app

romitg2 reviewed (commented):
@nangelina one question, otherwise seems good
